### PR TITLE
Add gpu proc info

### PR DESCRIPTION
**Pull request by jamartin9**
## Description

Consolidated gpu calls into `data_harvester`.

Changed config flag from `enable_gpu_memory` to `enable_gpu`.

~Added GPU utilization to the cpu widget.~

Added GPU process memory usage and utilization percentage to the proc widget. (enable by uncommenting the process column headers in the config file; not shown by default)
Added key binds for gpu process toggling `C` `M` (better defaults?)

~Added GPU power usage to the battery widget. (linux only as windows returned 0 for me.)~
Added bounds check to battery widget header (selected out of bounds if clicking across the top of battery widget from left->right; regardless of tabs).
~Show battery widget header (for gpu name) when `enable_gpu`. (does gpu power usage belong in the battery widget?)~
Added battery config/cli precedence 

Added feature flag `legacy-functions` to `nvml-wrapper`

updated config file(s)
updated help text
updated docs

Screenshots:

linux basic:

linux-basic

linux:

linux

freebsd:

freebsd

windows basic:

win-basic

## Issues

Maybe closes: #566 and #1291(disable temp with gpu flag) ~and #298 unless more information is desired (like fan speed for 272). Should there be a list of things still needed? (amd/intel gpus?)~

NVML seems inconsistent at times. Linux/Windows results differed after driver updates. (maybe keep `legacy-functions` feature until issue 48 resolved in nvml-wrapper...)

## Testing

Ran on windows(w/gpu), linux (w/gpu) and freebsd (no gpu)

- [X] _Windows_
- [ ] _macOS_
- [X] _Linux_
- [X] _FreeBSD_

## Checklist

- [X] _Areas your change affects have been linted using rustfmt (`cargo fmt`)_
- [X] _The change has been tested and doesn't appear to cause any unintended breakage_
- [X] _Documentation has been added/updated if needed (`README.md`, help menu, doc pages, etc.)_
- [X] _The pull request passes the provided CI pipeline_
- [X] _There are no merge conflicts_
- [ ] _If relevant, new tests were added (don't worry too much about coverage)_

I did *NOT* test on macOS.

If any of the Issues should be addressed or a different approach used please let me know.

Feel free to close this PR if it is unwanted.

### CODE REVIEW

File: src/app/data_harvester.rs
Comment (ClementTsang) [inline src/app/data_harvester.rs:382]: Hm, not sure how I feel about this part - power consumption is an interesting stat for sure, but I feel like sticking it as a battery metric feels a bit like a wrong move? Might be better as a new widget type or something.

File: docs/content/usage/widgets/process.md
Comment (ClementTsang) [inline docs/content/usage/widgets/process.md:156]: `gmem%` might also be good, though I can add this after.

File: src/app/data_harvester.rs
Comment (ClementTsang) [inline src/app/data_harvester.rs:325]: If we do use this big function for it, it might be more convenient to just pass a reference to `self.widgets_to_harvest`.

File: src/app/data_harvester.rs
Comment (ClementTsang) [inline src/app/data_harvester.rs:310]: Not sure if I would prefer doing it like this or delegating out some of these parts to the individual functions that already handle them.

File: src/app/data_harvester.rs
Comment (ClementTsang) [inline src/app/data_harvester.rs:354]: Not really a fan of this, again, probably better off putting this as a new widget, rather than combining data sources. Apologies if doing that is really hard/confusing, been looking at making new widgets easier...

File: src/app/query.rs
Comment (ClementTsang) [inline src/app/query.rs:635]: Should this be uncommented?

**Review by ClementTsang — CHANGES_REQUESTED** [review]
Sorry for taking so long - been busy with some other stuff.

Generally, I like what is done and do want to merge this; adding GPU support has been something I've wanted for a while. However, I feel like there's a few areas where I would rather GPU information be displayed (at least by default) as a separate widget, like usage and power consumption.

Feel free to let me know if there's any areas you would like to see changes in the code to make that easier.

File: src/app/data_harvester.rs
Comment (jamartin9) [reply src/app/data_harvester.rs:382]: I am fine with removing it. It can be added in a different widget later...

File: docs/content/usage/widgets/process.md
Comment (jamartin9) [reply docs/content/usage/widgets/process.md:156]: I can look into adding it into the mem% toggle too

File: src/app/data_harvester.rs
Comment (jamartin9) [reply src/app/data_harvester.rs:310]: Consolidation was to avoid iterating over the devices multiple times. ie the multiple `device_by_index` call loops that would be done for each collector. Might not be as significant if the power/util is removed.

File: src/app/data_harvester.rs
Comment (jamartin9) [reply src/app/data_harvester.rs:354]: I am fine with removing it. It can be added in a different widget later...

File: src/app/query.rs
Comment (jamartin9) [reply src/app/query.rs:635]: it is used in the above match arm. I can look into enabling the feature gate by pulling the above match arm branch into a function.

**Comment by jamartin9** [comment]
Thank you for your comments and feedback thus far.

Please take your time and apologizes for the large PR.

> there's a few areas where I would rather GPU information be displayed (at least by default) as a separate widget, like usage and power consumption.

With removal of the power/util the only added information is for procs... Should this be closed and I just create a PR for gpu proc info?

>Feel free to let me know if there's any areas you would like to see changes in the code to make that easier.

I have not looked into adding widgets yet. My initial concern is screen space in advanced mode (Thus why I added to existing displays).

**Comment by ClementTsang** [comment]
Sorry for the delay with reviewing this btw, been really with work/life for a bit - will get to looking at this in a bit.

**Comment by 0re5ama** [comment]
Any Progress on this?

**Comment by ClementTsang** [comment]
Sorry, just got back from a vacation 😅 , will look at this sometime this week.

**Review by ClementTsang — COMMENTED** [review]
Looks mostly good to me, apologies for taking so long to review! Just a few comments/questions/nitpicks but bulk of it seems fine.

File: docs/content/configuration/config-file/processes.md
Comment (jamartin9) [reply docs/content/configuration/config-file/processes.md]: I would appreciate it if you could add to the docs to reflect that the column values will not be populated without `enable-gpu`.

File: src/app.rs
Comment (jamartin9) [reply src/app.rs:1303]: If the columns are enabled it sorts via the empty values. Without the columns it returns `None` as the column index does not exist.

File: src/app/data_harvester.rs
Comment (jamartin9) [reply src/app/data_harvester.rs]: done

File: src/app/data_harvester.rs
Comment (jamartin9) [reply src/app/data_harvester.rs:313]: done

File: src/app/data_harvester/nvidia.rs
Comment (jamartin9) [reply src/app/data_harvester/nvidia.rs]: done

File: src/app/data_harvester/processes/linux.rs
Comment (jamartin9) [reply src/app/data_harvester/processes/linux.rs]: done

File: src/args.rs
Comment (jamartin9) [reply src/args.rs]: done

File: src/widgets/process_table/proc_widget_column.rs
Comment (jamartin9) [reply src/widgets/process_table/proc_widget_column.rs]: done

File: src/options.rs
Comment (jamartin9) [reply src/options.rs:759]: > why this change?

The change was made for the 'battery config/cli precedence' to show the widget regardless of detected batteries. I was using it for the gpu power and the tab layout bounds when without a battery. 

This was a future change suggested in the PR that ignored the battery widget without detected batteries. 

> #975 
"If needed, it can be adjusted so that if the config file contains the battery field, it will still show the widget even without batteries on the system."

The suggested change:
> more like:
`if use_battery {
  if let Ok(battery_manager) = Manager::new() {
    // ...
  }
} else {
  false
}`

would again ignore the no battery present but still want the widget case as the commented out code above is
`if batteries.count() == 0 {
                    return false;
}`

File: src/app/data_harvester/temperature.rs
Comment (jamartin9) [reply src/app/data_harvester/temperature.rs]: done

File: src/app/data_harvester.rs
Comment (jamartin9) [reply src/app/data_harvester.rs]: removed.

File: src/widgets/process_table/proc_widget_column.rs
Comment (ClementTsang) [inline src/widgets/process_table/proc_widget_column.rs:96]: Nit for consistency:

```suggestion
            ProcColumn::GpuMem => "GMem",
            #[cfg(feature = "gpu")]
            ProcColumn::GpuMemPercent => "GMem%",
```

File: src/widgets/process_table/proc_widget_column.rs
Comment (ClementTsang) [inline src/widgets/process_table/proc_widget_column.rs:122]: Same here:

```suggestion
            #[cfg(feature = "gpu")]
            ProcColumn::GpuMem => "GMem",
            #[cfg(feature = "gpu")]
            ProcColumn::GpuMemPercent => "GMem%",
```

File: src/options.rs
Comment (ClementTsang) [inline src/options.rs:436]: Can probably just do:

```suggestion
        use_gpu: get_enable_gpu(matches, config),
```

**Review by ClementTsang — APPROVED** [review]
Small nitpicks, but otherwise looks good to me. Happy to merge if you don't have anything else to add in.

**Comment by jamartin9** [comment]
> Small nitpicks, but otherwise looks good to me. Happy to merge if you don't have anything else to add in.

I do not have anything to add unless you want me to address the comments. I can assist in testing after the merge for the changes too.

Thanks again for the responses.

**Comment by ClementTsang** [comment]
They're small things that I can address quickly after merging this, so no worries there.

And testing would be appreciated! I don't have any computers that use NVIDIA GPUs at the moment so it's a bit hard for me to test right now unless I mock it or until I look into supporting other GPUs.

**Comment by actionless** [comment]
all widgets (like processes, memory or temp) work as expected but on this one i'm not getting `GPU0` etc, do i need to enable some config option for that?

2024-01-02--1704177666_2244x496_scrot

**Comment by jamartin9** [comment]
> all widgets (like processes, memory or temp) work as expected but on this one i'm not getting `GPU0` etc, do i need to enable some config option for that?

It was determined during code review ([commit](https://github.com/ClementTsang/bottom/pull/1276/commits/4234ec233b39c9f3840e4d9e6e39ed2956d3f7c0)) to remove the utilization and power consumption from the default displays per the following comment:

>  I would rather GPU information be displayed as a separate widget, like usage and power consumption.

I didn't update the screenshots after removal. Sorry for the confusion.

**Comment by actionless** [comment]
and that separate widget is implemented or not yet?

**Comment by actionless** [comment]
btw i also find it a bit confusing showing VRAM and RAM memory but split GPU and CPU processors - in that case i guess VRAM also should be moved separately?

**Comment by jamartin9** [comment]
> and that separate widget is implemented or not yet?

No it is not. Anyone is free to implement it tho.

> VRAM also should be moved to the gpu widget?

Most likely imo

**Pull request merged**
